### PR TITLE
[TRUNK-344] PersonName.getFullName() should use the configured nameLayout

### PR DESCRIPTION
**Pull request by BartlomiejRasztabiga**
Issue: https://issues.openmrs.org/browse/TRUNK-344

I've made also few changes to NameSupport so it may be reusable and pass all tests

### CODE REVIEW

**Comment by BartlomiejRasztabiga** [comment]
In addition, I think we can replace call to NameSupport constructor with getInstance method, because Singleton shouldn't expose its' constructor publicly.

File: api/src/main/java/org/openmrs/layout/name/NameSupport.java [hunk 1]
```diff
@@ -23,18 +23,15 @@
...
+	private NameSupport() {
+		// hiding default public constructor
 	}
 	
 	public static NameSupport getInstance() {
 		if (singleton == null) {
-			throw new RuntimeException("Not Yet Instantiated");
-		} else {
-			return singleton;
+			singleton = new NameSupport();
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/layout/name/NameSupport.java]: Isn't this resulting into a bug caused by a bean different from this? https://github.com/openmrs/openmrs-module-legacyui/blob/master/omod/src/main/resources/webModuleApplicationContext.xml#L754

**Comment by BartlomiejRasztabiga** [comment]
So we have to initialise NameSupport instance before calling getInstance() because otherwise it doesn't pass tests.

**Comment by BartlomiejRasztabiga** [comment]
So If I understand correctly, during normal application run, NameSupport is created automatically by Spring?

**Comment by dkayiwa** [comment]
Correct!

**Comment by BartlomiejRasztabiga** [comment]
I have an issue. If I call NameSupport.getInstance() in getFullName() then I get "Not Yet Instantiated" exception in every test related to it. So we have to initialize this Singleton first, right?

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa I think I've fixed it :)

File: api/src/main/java/org/openmrs/layout/name/NameSupport.java [hunk 2]
```diff
@@ -36,6 +36,12 @@ public static NameSupport getInstance() {
 			return singleton;
 		}
 	}
+
+	public static void init() {
+		if (singleton == null) {
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/layout/name/NameSupport.java]: Is this just for testing?

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: This method is used to initialise first instance of NameSupport, otherwise it will throw Not Yet Instantiated"

File: api/src/main/java/org/openmrs/PersonName.java [hunk 3]
```diff
@@ -11,6 +11,7 @@
 
 import static org.apache.commons.lang.StringUtils.defaultString;
 
+import java.lang.reflect.Method;
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/PersonName.java]: Where is this import used?

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/PersonName.java]: I can remove that :)

File: api/src/main/java/org/openmrs/layout/name/NameSupport.java (hunk 2 above)
Comment (dkayiwa) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: Let me rephrase, did you add the init method simply for tests?

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: I think so, because Spring doesn't initialise this bean only in unit tests :blush:

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: Did you consider putting the bean in TestingApplicationContext.xml? 😊

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: So if I add something like this into TestingApplicationContext it would autowire NameTemplate?

        ```<bean id="nameSupport"
	      class="org.openmrs.layout.name.NameSupport">
	```</bean>

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: @dkayiwa Could you tell me how to initialise this class instance before running any test? 😊

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: I've put something like this
https://github.com/openmrs/openmrs-core/pull/2470/files#diff-de63ef1337d1da40a8826bb63ee8fc92R33

Comment (dkayiwa) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: And did it work? 😊

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/layout/name/NameSupport.java]: Unfortunately no, that's why I ask you

File: api/src/test/resources/TestingApplicationContext.xml [hunk 4]
```diff
@@ -30,6 +30,11 @@
 		</property>
 	</bean>
 
+	<bean id="nameSupport"
+	      class="org.openmrs.layout.name.NameSupport"
+	      factory-method="getInstance">
```
Comment (dkayiwa) [inline api/src/test/resources/TestingApplicationContext.xml]: You can remove factory-method

File: api/src/test/java/org/openmrs/layout/name/NameTemplateTest.java [hunk 5]
```diff
@@ -27,7 +27,8 @@
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/layout/name/NameTemplateTest.java]: Change this test's base class to BaseContextSensitiveTest

Comment (dkayiwa) [reply api/src/test/java/org/openmrs/layout/name/NameTemplateTest.java]: And in BaseContextSensitiveTest add "classpath*:TestingApplicationContext.xml"

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa Thanks! I've made appropiate changes 😊
